packages/typespec-client-generator-core/src/validate.ts

```diff
@@ -1,12 +1,177 @@
+  for (const model of namespace.models.values()) {
+    validateClientNamesCore(program, tcgcContext, scope, model.properties.values());
+  }
+
+  // Check for duplicate client names for enum members
+  for (const item of namespace.enums.values()) {
+    validateClientNamesCore(program, tcgcContext, scope, item.members.values());
+  }
+
+  // Check for duplicate client names for union variants
+  for (const item of namespace.unions.values()) {
+    validateClientNamesCore(program, tcgcContext, scope, item.variants.values());
+  }
+
+  // Check for duplicate client names for nested namespaces
+  for (const item of namespace.namespaces.values()) {
+    validateClientNamesPerNamespace(program, tcgcContext, scope, item);
+  }
+}
+
+function validateClientNamesCore(
+  program: Program,
+  tcgcContext: TCGCContext,
+  scope: string | symbol,
+  items: Iterable<
+    | Namespace
+    | Scalar
+    | Operation
+    | Interface
+    | Model
+    | Enum
+    | Union
+    | ModelProperty
+    | EnumMember
+    | UnionVariant
+  >
+) {
+  const duplicateTracker = new DuplicateTracker<string, Type>();
+
+  for (const item of items) {
+    const clientName = getClientNameOverride(tcgcContext, item, scope);
+    if (clientName !== undefined) {
+      const clientNameDecorator = item.decorators.find((x) => x.definition?.name === "@clientName");
+      if (clientNameDecorator?.definition !== undefined) {
+        duplicateTracker.track(clientName, clientNameDecorator.definition);
+      }
+    } else {
+      if (item.name !== undefined && typeof item.name === "string") {
+        duplicateTracker.track(item.name, item);
+      }
+    }
+  }
+
+  reportDuplicateClientNames(program, duplicateTracker, scope);
+}
+
+function reportDuplicateClientNames(
+  program: Program,
+  duplicateTracker: DuplicateTracker<string, Type>,
+  scope: string | symbol
+) {
+  for (const [name, duplicates] of duplicateTracker.entries()) {
+    for (const item of duplicates) {
+      if (scope === AllScopes) {
+        if (item.kind === "Decorator") {
+          reportDiagnostic(program, {
+            code: "duplicate-client-name",
+            format: { name, scope: "AllScopes" },
+            target: item,
+          });
+        } else {
+          reportDiagnostic(program, {
+            code: "duplicate-client-name",
+            messageId: "nonDecorator",
+            format: { name, scope: "AllScopes" },
+            target: item,
+          });
+        }
+      } else if (typeof scope === "string") {
+        if (item.kind === "Decorator") {
+          reportDiagnostic(program, {
+            code: "duplicate-client-name",
+            format: { name, scope },
+            target: item,
+          });
+        } else {
+          reportDiagnostic(program, {
+            code: "duplicate-client-name",
+            messageId: "nonDecorator",
+            format: { name, scope },
+            target: item,
+          });
```

Comment:
```suggestion
      const scopeStr = scope === AllScopes ? "AllScopes" : scope;
      if (item.kind === "Decorator") {
        reportDiagnostic(program, {
          code: "duplicate-client-name",
          format: { name, scope: scopeStr },
          target: item,
        });
      } else {
        reportDiagnostic(program, {
          code: "duplicate-client-name",
          messageId: "nonDecorator",
          format: { name, scope: scopeStr },
          target: item,
        });
```
can remove duplication here i think

Reply:
```
const scopeStr = scope === AllScopes ? "AllScopes" : scope;
```
I tried this before, this doesn't really work.
The type of `scopeStr` here is `string | symbol`, we can't simply put it into for message.

```
const scopeStr = typeof scope === "string" ? scope : "AllScopes";
```
This could work since it ensures the type is `string`, but it will assign `AllScopes` under unexpected conditions.
So, I would prefer to keep the redundant code to make things work properly.

Reply:
are you sure you did that correctly? this works fine for me with a symbol

<img width="422" alt="image" src="https://github.com/user-attachments/assets/b83490e1-27f9-40ff-b788-2703ac4dff36">

this is just a **lot** of duplicate code

Reply:
OK, it's safe to case `string | symbol` to `string` in this case anyway, updated it.

Reply:
It’s but the code should always be a string here and typescript is smart enough to understand that

Reply:
Apparently, typescript is not smart enough to understand this if I remove the casting, got error as below
```
src/validate.ts:146:27 - error TS2322: Type 'string | symbol' is not assignable to type 'string'.
```

Reply:
So as shown above it works, please check you have the right type to start with

Reply:
If the type is symbol | string then this whole think is wrong anyway

Reply:
The type should be typeof AllScopes | string

Reply:
`AllScopes` is a value of `Symbol`, not a type though

Reply:
That’s why there is a typeof in front, please compare with my screenshot above

Reply:
updated